File: scripts/train_fast.py
#!/usr/bin/env python3
"""
FAST Training Pipeline - Parquet + DuckDB + Parallel

Uses:
1. Parquet files for 10-50x faster reads
2. DuckDB for instant aggregations
3. Pre-computed matched pairs (saved to disk)
4. Parallel training across entities
"""

import argparse
import json
import logging
import sys
from datetime import datetime
from pathlib import Path
from concurrent.futures import ProcessPoolExecutor, as_completed
import multiprocessing

import numpy as np
import pandas as pd
import duckdb
from zoneinfo import ZoneInfo

# Add src to path
if str(Path(__file__).parent.parent / "src") not in sys.path:
    sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

try:
    import xgboost as xgb
except ImportError:
    xgb = None

# Constants
MATCH_WINDOW_MINUTES = 15
DEFAULT_MIN_OBS = 500
DEFAULT_FALLBACK_RATIO = 0.82
DEFAULT_WORKERS = 8

PREDICTOR_COLUMNS = [
    "mins_since_6am",
    "hour_of_day",
    "day_of_week",
    "month",
    "is_weekend",
]


def setup_logging(output_base: Path):
    log_dir = output_base / "logs"
    log_dir.mkdir(parents=True, exist_ok=True)
    
    log_file = log_dir / f"train_fast_{datetime.now().strftime('%Y%m%d_%H%M%S')}.log"
    
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s [%(levelname)s] %(message)s",
        handlers=[
            logging.FileHandler(log_file, encoding="utf-8"),
            logging.StreamHandler(sys.stdout),
        ],
    )
    return logging.getLogger(__name__)


def get_entity_counts_duckdb(parquet_dir: Path, logger) -> dict:
    """Use DuckDB to instantly count ACTUAL observations per entity."""
    logger.info("Counting ACTUAL observations with DuckDB (fast!)...")
    
    con = duckdb.connect()
    
    result = con.execute(f"""
        SELECT 
            entity_code,
            COUNT(*) as actual_count
        FROM read_parquet('{parquet_dir}/*.parquet')
        WHERE wait_time_type = 'ACTUAL'
        GROUP BY entity_code
        ORDER BY actual_count DESC
    """).fetchdf()
    
    counts = dict(zip(result["entity_code"], result["actual_count"]))
    logger.info(f"  Found {len(counts)} entities with ACTUAL data")
    
    con.close()
    return counts


def create_all_matched_pairs_duckdb(parquet_dir: Path, output_path: Path, logger) -> int:
    """Use DuckDB to create all matched pairs at once (vectorized, fast!)."""
    logger.info("Creating matched pairs with DuckDB (vectorized, fast!)...")
    
    con = duckdb.connect()
    
    # This query matches ACTUAL with closest POSTED within 15-minute window
    query = f"""
        WITH actual AS (
            SELECT 
                entity_code,
                observed_at,
                observed_at_ts,
                park_date,
                wait_time_minutes as actual_time
            FROM read_parquet('{parquet_dir}/*.parquet')
            WHERE wait_time_type = 'ACTUAL'
              AND wait_time_minutes IS NOT NULL
              AND wait_time_minutes > 0
        ),
        posted AS (
            SELECT 
                entity_code,
                observed_at_ts,
                park_date,
                wait_time_minutes as posted_time
            FROM read_parquet('{parquet_dir}/*.parquet')
            WHERE wait_time_type = 'POSTED'
              AND wait_time_minutes IS NOT NULL
              AND wait_time_minutes > 0
        ),
        matched AS (
            SELECT 
                a.entity_code,
                a.observed_at,
                a.observed_at_ts,
                a.park_date,
                a.actual_time,
                p.posted_time,
                ABS(EXTRACT(EPOCH FROM (a.observed_at_ts - p.observed_at_ts))) as time_diff_sec
            FROM actual a
            JOIN posted p 
              ON a.entity_code = p.entity_code 
              AND a.park_date = p.park_date
              AND ABS(EXTRACT(EPOCH FROM (a.observed_at_ts - p.observed_at_ts))) <= {MATCH_WINDOW_MINUTES * 60}
        ),
        best_match AS (
            SELECT 
                entity_code,
                observed_at,
                observed_at_ts,
                park_date,
                actual_time,
                posted_time,
                ROW_NUMBER() OVER (
                    PARTITION BY entity_code, observed_at 
                    ORDER BY time_diff_sec
                ) as rn
            FROM matched
        )
        SELECT 
            entity_code,
            observed_at,
            observed_at_ts,
            park_date,
            actual_time,
            posted_time
        FROM best_match
        WHERE rn = 1
    """
    
    # Execute and save to parquet
    logger.info("  Running match query...")
    df = con.execute(query).fetchdf()
    
    logger.info(f"  Created {len(df):,} matched pairs")
    
    # Add simple features
    df["observed_at_ts"] = pd.to_datetime(df["observed_at_ts"])
    df["hour_of_day"] = df["observed_at_ts"].dt.hour
    df["mins_since_6am"] = (df["observed_at_ts"].dt.hour - 6) * 60 + df["observed_at_ts"].dt.minute
    df["day_of_week"] = df["observed_at_ts"].dt.dayofweek
    df["month"] = df["observed_at_ts"].dt.month
    df["is_weekend"] = (df["day_of_week"] >= 5).astype(int)
    
    # Save to parquet
    output_path.parent.mkdir(parents=True, exist_ok=True)
    df.to_parquet(output_path, index=False)
    logger.info(f"  Saved matched pairs to: {output_path}")
    
    con.close()
    return len(df)


def train_single_entity(args):
    """Train model for a single entity (called in parallel)."""
    entity_code, matched_pairs_path, models_dir, min_samples = args
    
    try:
        # Load matched pairs for this entity
        df = pd.read_parquet(
            matched_pairs_path,
            filters=[("entity_code", "==", entity_code)]
        )
        
        if len(df) < min_samples:
            return entity_code, None, f"Not enough samples ({len(df)})"
        
        # Prepare features
        feature_cols = ["posted_time"] + [c for c in PREDICTOR_COLUMNS if c in df.columns]
        X = df[feature_cols].fillna(-1)
        y = df["actual_time"]
        
        # Remove invalid rows
        valid = y.notna() & (y > 0)
        X = X[valid]
        y = y[valid]
        
        if len(X) < min_samples:
            return entity_code, None, f"Not enough valid samples ({len(X)})"
        
        # Train/val split (chronological)
        n = len(X)
        train_end = int(n * 0.85)
        
        X_train, y_train = X.iloc[:train_end], y.iloc[:train_end]
        X_val, y_val = X.iloc[train_end:], y.iloc[train_end:]
        
        # Train XGBoost
        model = xgb.XGBRegressor(
            n_estimators=500,
            max_depth=6,
            learning_rate=0.1,
            early_stopping_rounds=20,
            random_state=42,
        )
        
        model.fit(X_train, y_train, eval_set=[(X_val, y_val)], verbose=False)
        
        # Evaluate
        y_pred = model.predict(X_val)
        mae = np.mean(np.abs(y_val - y_pred))
        
        # Save model
        model_dir = models_dir / entity_code
        model_dir.mkdir(parents=True, exist_ok=True)
        model.save_model(str(model_dir / "model.json"))
        
        # Save metadata
        metadata = {
            "entity_code": entity_code,
            "trained_at": datetime.now(ZoneInfo("UTC")).isoformat(),
            "n_samples": len(X_train),
            "mae": float(mae),
            "features": feature_cols,
        }
        with open(model_dir / "metadata.json", "w") as f:
            json.dump(metadata, f, indent=2)
        
        return entity_code, len(X_train), mae
    
    except Exception as e:
        return entity_code, None, str(e)


def main():
    parser = argparse.ArgumentParser(description="Fast training pipeline")
    parser.add_argument("--parquet-dir", default="/home/wilma/hazeydata/pipeline/fact_tables/parquet")
    parser.add_argument("--output-base", default="/home/wilma/hazeydata/pipeline")
    parser.add_argument("--min-obs", type=int, default=DEFAULT_MIN_OBS)
    parser.add_argument("--workers", type=int, default=DEFAULT_WORKERS)
    parser.add_argument("--skip-matching", action="store_true", help="Use existing matched pairs")
    
    args = parser.parse_args()
    
    if xgb is None:
        print("ERROR: XGBoost not installed")
        sys.exit(1)
    
    parquet_dir = Path(args.parquet_dir)
    output_base = Path(args.output_base)
    models_dir = output_base / "models"
    matched_pairs_path = output_base / "matched_pairs" / "all_pairs.parquet"
    
    logger = setup_logging(output_base)
    
    logger.info("=" * 60)
    logger.info("FAST TRAINING PIPELINE")
    logger.info("=" * 60)
    logger.info(f"Parquet dir: {parquet_dir}")
    logger.info(f"Min observations: {args.min_obs}")
    logger.info(f"Workers: {args.workers}")
    
    # Step 1: Get entity counts
    start = datetime.now()
    entity_counts = get_entity_counts_duckdb(parquet_dir, logger)
    logger.info(f"  Counting took {(datetime.now() - start).total_seconds():.1f}s")
    
    # Split by threshold
    entities_to_train = [e for e, c in entity_counts.items() if c >= args.min_obs]
    entities_fallback = [e for e, c in entity_counts.items() if c < args.min_obs]
    
    logger.info(f"\nEntities with >= {args.min_obs} ACTUAL: {len(entities_to_train)}")
    logger.info(f"Entities with < {args.min_obs} ACTUAL: {len(entities_fallback)} (82% ratio)")
    
    # Show top entities
    sorted_counts = sorted(entity_counts.items(), key=lambda x: x[1], reverse=True)
    logger.info(f"\nTop 20 entities:")
    for e, c in sorted_counts[:20]:
        logger.info(f"  {e}: {c:,}")
    
    # Step 2: Create matched pairs (if needed)
    if not args.skip_matching or not matched_pairs_path.exists():
        start = datetime.now()
        n_pairs = create_all_matched_pairs_duckdb(parquet_dir, matched_pairs_path, logger)
        logger.info(f"  Matching took {(datetime.now() - start).total_seconds():.1f}s")
    else:
        logger.info(f"\nUsing existing matched pairs: {matched_pairs_path}")
    
    # Step 3: Train models in parallel
    logger.info(f"\n{'='*60}")
    logger.info(f"TRAINING {len(entities_to_train)} ENTITY MODELS (parallel)")
    logger.info(f"{'='*60}")
    
    start = datetime.now()
    
    work_items = [
        (entity, matched_pairs_path, models_dir, 100)
        for entity in entities_to_train
    ]
    
    successful = 0
    failed = 0
    
    with ProcessPoolExecutor(max_workers=args.workers) as executor:
        futures = {executor.submit(train_single_entity, item): item[0] for item in work_items}
        
        for future in as_completed(futures):
            entity = futures[future]
            result = future.result()
            
            if result[1] is not None:
                successful += 1
                if successful % 20 == 0:
                    logger.info(f"  Trained {successful}/{len(entities_to_train)} models...")
            else:
                failed += 1
                logger.warning(f"  {entity}: {result[2]}")
    
    elapsed = (datetime.now() - start).total_seconds()
    
    logger.info(f"\n{'='*60}")
    logger.info(f"TRAINING COMPLETE")
    logger.info(f"{'='*60}")
    logger.info(f"Successful: {successful}")
    logger.info(f"Failed: {failed}")
    logger.info(f"Training time: {elapsed:.1f}s ({elapsed/len(entities_to_train):.2f}s per entity)")
    logger.info(f"Models saved to: {models_dir}")


if __name__ == "__main__":
    main()

File: julia-ml/train_fast.jl
#!/usr/bin/env julia
"""
FAST Training Pipeline - Julia Version

Benchmarking Julia vs Python for XGBoost training.
Uses Parquet files, native threading, and XGBoost.jl.
"""

using Dates
using DataFrames
using Parquet2
using JSON3
using XGBoost
using Statistics

# Constants
const MATCH_WINDOW_MINUTES = 15
const DEFAULT_MIN_OBS = 500
const DEFAULT_FALLBACK_RATIO = 0.82

const PREDICTOR_COLUMNS = [
    :mins_since_6am,
    :hour_of_day,
    :day_of_week,
    :month,
    :is_weekend,
]

"""Load all parquet files from directory into a single DataFrame."""
function load_parquet_dir(dir::String)
    files = filter(f -> endswith(f, ".parquet"), readdir(dir, join=true))
    println("  Loading $(length(files)) parquet files...")
    
    dfs = DataFrame[]
    for f in files
        push!(dfs, DataFrame(Parquet2.Dataset(f)))
    end
    
    return vcat(dfs...)
end

"""Count ACTUAL observations per entity."""
function get_entity_counts(df::DataFrame)
    println("Counting ACTUAL observations...")
    
    actual_df = filter(row -> row.wait_time_type == "ACTUAL", df)
    counts = combine(groupby(actual_df, :entity_code), nrow => :actual_count)
    
    println("  Found $(nrow(counts)) entities with ACTUAL data")
    return Dict(row.entity_code => row.actual_count for row in eachrow(counts))
end

"""Create matched pairs (ACTUAL with closest POSTED within window)."""
function create_matched_pairs(df::DataFrame, output_path::String)
    println("Creating matched pairs...")
    
    # Split by type
    actual = filter(row -> row.wait_time_type == "ACTUAL" && 
                          !ismissing(row.wait_time_minutes) && 
                          row.wait_time_minutes > 0, df)
    posted = filter(row -> row.wait_time_type == "POSTED" && 
                          !ismissing(row.wait_time_minutes) && 
                          row.wait_time_minutes > 0, df)
    
    println("  ACTUAL rows: $(nrow(actual))")
    println("  POSTED rows: $(nrow(posted))")
    
    # Create lookup by (entity, park_date)
    posted_lookup = Dict{Tuple{String, Any}, Vector{NamedTuple}}()
    for row in eachrow(posted)
        key = (row.entity_code, row.park_date)
        if !haskey(posted_lookup, key)
            posted_lookup[key] = []
        end
        push!(posted_lookup[key], (ts=row.observed_at_ts, posted=row.wait_time_minutes))
    end
    
    # Match each ACTUAL with closest POSTED
    matched = NamedTuple[]
    window_sec = MATCH_WINDOW_MINUTES * 60
    
    for (i, row) in enumerate(eachrow(actual))
        if i % 100000 == 0
            println("  Processed $i / $(nrow(actual)) ACTUAL rows...")
        end
        
        key = (row.entity_code, row.park_date)
        candidates = get(posted_lookup, key, [])
        
        best_posted = nothing
        best_diff = Inf
        
        actual_ts = row.observed_at_ts
        
        for c in candidates
            diff = abs(Dates.value(actual_ts - c.ts) / 1000)  # milliseconds to seconds
            if diff <= window_sec && diff < best_diff
                best_diff = diff
                best_posted = c.posted
            end
        end
        
        if best_posted !== nothing
            push!(matched, (
                entity_code = row.entity_code,
                observed_at = row.observed_at,
                observed_at_ts = actual_ts,
                park_date = row.park_date,
                actual_time = row.wait_time_minutes,
                posted_time = best_posted,
            ))
        end
    end
    
    println("  Created $(length(matched)) matched pairs")
    
    # Convert to DataFrame and add features
    result = DataFrame(matched)
    
    result.hour_of_day = hour.(result.observed_at_ts)
    result.mins_since_6am = (hour.(result.observed_at_ts) .- 6) .* 60 .+ minute.(result.observed_at_ts)
    result.day_of_week = dayofweek.(result.observed_at_ts)
    result.month = month.(result.observed_at_ts)
    result.is_weekend = Int.(dayofweek.(result.observed_at_ts) .>= 6)
    
    # Save to parquet
    mkpath(dirname(output_path))
    Parquet2.writefile(output_path, result)
    println("  Saved to: $output_path")
    
    return result
end

"""Train XGBoost model for a single entity."""
function train_single_entity(entity_code::String, matched_df::DataFrame, models_dir::String, min_samples::Int)
    # Filter for this entity
    entity_df = filter(row -> row.entity_code == entity_code, matched_df)
    
    if nrow(entity_df) < min_samples
        return nothing, "Not enough samples ($(nrow(entity_df)))"
    end
    
    # Prepare features
    feature_cols = [:posted_time, :mins_since_6am, :hour_of_day, :day_of_week, :month, :is_weekend]
    X = Matrix{Float32}(entity_df[:, feature_cols])
    y = Float32.(entity_df.actual_time)
    
    # Remove invalid
    valid = .!isnan.(y) .& (y .> 0)
    X = X[valid, :]
    y = y[valid]
    
    if length(y) < min_samples
        return nothing, "Not enough valid samples ($(length(y)))"
    end
    
    # Train/val split (chronological)
    n = length(y)
    train_end = floor(Int, n * 0.85)
    
    X_train, y_train = X[1:train_end, :], y[1:train_end]
    X_val, y_val = X[train_end+1:end, :], y[train_end+1:end]
    
    # Train XGBoost
    dtrain = DMatrix(X_train, label=y_train)
    dval = DMatrix(X_val, label=y_val)
    
    params = Dict(
        "max_depth" => 6,
        "eta" => 0.1,
        "objective" => "reg:squarederror",
        "seed" => 42,
    )
    
    watchlist = [(dtrain, "train"), (dval, "eval")]
    
    bst = xgboost(dtrain, 500;
                  watchlist=watchlist,
                  params...,
                  early_stopping_rounds=20,
                  verbosity=0)
    
    # Evaluate
    y_pred = XGBoost.predict(bst, dval)
    mae = mean(abs.(y_val .- y_pred))
    
    # Save model
    model_dir = joinpath(models_dir, entity_code)
    mkpath(model_dir)
    XGBoost.save(bst, joinpath(model_dir, "model_julia.json"))
    
    # Save metadata
    metadata = Dict(
        "entity_code" => entity_code,
        "trained_at" => Dates.format(now(Dates.UTC), "yyyy-mm-ddTHH:MM:SS"),
        "n_samples" => train_end,
        "mae" => mae,
        "features" => string.(feature_cols),
        "backend" => "Julia XGBoost.jl",
    )
    open(joinpath(model_dir, "metadata_julia.json"), "w") do f
        JSON3.write(f, metadata)
    end
    
    return train_end, mae
end

"""Main training pipeline."""
function main(; parquet_dir::String="/home/wilma/hazeydata/pipeline/fact_tables/parquet",
               output_base::String="/home/wilma/hazeydata/pipeline",
               min_obs::Int=DEFAULT_MIN_OBS,
               skip_matching::Bool=false)
    
    models_dir = joinpath(output_base, "models")
    matched_pairs_path = joinpath(output_base, "matched_pairs", "all_pairs_julia.parquet")
    
    println("=" ^ 60)
    println("FAST TRAINING PIPELINE (Julia)")
    println("=" ^ 60)
    println("Parquet dir: $parquet_dir")
    println("Min observations: $min_obs")
    println("Threads: $(Threads.nthreads())")
    
    # Load all data
    println("\nLoading data...")
    load_start = time()
    df = load_parquet_dir(parquet_dir)
    println("  Loaded $(nrow(df)) rows in $(round(time() - load_start, digits=1))s")
    
    # Get entity counts
    count_start = time()
    entity_counts = get_entity_counts(df)
    println("  Counting took $(round(time() - count_start, digits=1))s")
    
    # Split by threshold
    entities_to_train = [e for (e, c) in entity_counts if c >= min_obs]
    entities_fallback = [e for (e, c) in entity_counts if c < min_obs]
    
    println("\nEntities with >= $min_obs ACTUAL: $(length(entities_to_train))")
    println("Entities with < $min_obs ACTUAL: $(length(entities_fallback)) (82% ratio)")
    
    # Create or load matched pairs
    local matched_df
    if !skip_matching || !isfile(matched_pairs_path)
        match_start = time()
        matched_df = create_matched_pairs(df, matched_pairs_path)
        println("  Matching took $(round(time() - match_start, digits=1))s")
    else
        println("\nLoading existing matched pairs...")
        matched_df = DataFrame(Parquet2.Dataset(matched_pairs_path))
    end
    
    # Train models
    println("\n" * "=" ^ 60)
    println("TRAINING $(length(entities_to_train)) ENTITY MODELS")
    println("=" ^ 60)
    
    train_start = time()
    successful = 0
    failed = 0
    
    # Note: Julia threading for XGBoost can be tricky
    # Using serial for stability, but XGBoost itself uses threads internally
    for (i, entity) in enumerate(entities_to_train)
        result, msg = train_single_entity(entity, matched_df, models_dir, 100)
        
        if result !== nothing
            successful += 1
        else
            failed += 1
            # println("  $entity: $msg")
        end
        
        if i % 20 == 0
            println("  Trained $i/$(length(entities_to_train)) models...")
        end
    end
    
    elapsed = time() - train_start
    
    println("\n" * "=" ^ 60)
    println("TRAINING COMPLETE")
    println("=" ^ 60)
    println("Successful: $successful")
    println("Failed: $failed")
    println("Training time: $(round(elapsed, digits=1))s ($(round(elapsed/length(entities_to_train), digits=2))s per entity)")
    println("Models saved to: $models_dir")
    
    return successful, elapsed
end

# Run if called directly
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
